Testing

ericgaspar opened the pull request:
## Problem

- *Description of why you made this PR*

## Solution

- *And how do you fix that problem*

## PR Status

- [ ] Code finished and ready to be reviewed/tested
- [ ] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

yunohost-bot committed:
Auto-update README

ericgaspar committed:
cleaning

ericgaspar committed:
Merge branch 'testing' of https://github.com/YunoHost-Apps/kbin_ynh into testing

ericgaspar commented:
!testme

ericgaspar committed:
Update nginx.conf

ericgaspar committed:
Update manifest.toml

ericgaspar committed:
Update manifest.toml

ericgaspar committed:
Update manifest.toml

ericgaspar committed:
cleaning

yunohost-bot committed:
Auto-update README

ericgaspar committed:
Update install

ericgaspar committed:
Merge branch 'testing' of https://github.com/YunoHost-Apps/kbin_ynh into testing

ericgaspar committed:
Update install

ericgaspar committed:
Update manifest.toml

ericgaspar committed:
Update install

ericgaspar committed:
cleaning

yunohost-bot committed:
Auto-update README

ericgaspar committed:
Update install

ericgaspar committed:
Merge branch 'testing' of https://github.com/YunoHost-Apps/kbin_ynh into testing

ericgaspar committed:
Update nginx.conf

ericgaspar committed:
Update manifest.toml

ericgaspar committed:
Update install

ericgaspar committed:
Update _common.sh

ericgaspar committed:
cleaning

ericgaspar committed:
Update .env

ericgaspar committed:
Update install

ericgaspar committed:
Update install

ericgaspar committed:
Update install

ericgaspar committed:
Update install

ericgaspar committed:
Update install

ericgaspar committed:
Update install

ericgaspar committed:
cleaning

ericgaspar committed:
Update manifest.toml

ericgaspar committed:
Update install

ericgaspar committed:
Update install

ericgaspar committed:
Update install

csolisr commented:
!testme

csolisr commented:
```
 Do not run Composer as root/super user! See https://getcomposer.org/root for details
Aborting as no plugin should be loaded if running as super user is not explicitly allowed
```

Odd, I thought `ynh_install_composer` automatically dealt with the privileges issue?

tituspijean commented:
It does not! 

https://yunohost.org/en/packaging_apps_helpers#ynh-install-composer and its "Dude, show me the code!" link: https://github.com/YunoHost/yunohost/blob/19eb48b6e73267685e1417122ac47908c1cf2472/helpers/php#L522

There is no `ynh_exec_as` or `sudo -u $app` in there.

csolisr commented:
Hey @ericgaspar , mind if I upload a patch to this branch to solve the `ynh_install_composer` issue (and continue the packaging), or would you rather do it yourself?

csolisr committed:
fix: reduce linter warnings

csolisr commented:
!testme

csolisr commented:
Sorry about that, I just found that `php-symfony-mercure` is only shipped on Debian 12 or higher.

csolisr committed:
fix: remove php-symfony-mercure dependency

csolisr commented:
Another snag we hit: turns out that I can't simply chain `ynh_exec_as` and `ynh_install_composer` - the latter doesn't get properly escaped. Any ideas?

```
101202 INFO INFO - [#######++++.........] > Installing kbin with Composer...
101202 INFO DEBUG - + ynh_exec_as kbin ynh_install_composer
101202 INFO DEBUG - + local user=kbin
101202 INFO DEBUG - + shift 1
101202 INFO DEBUG - ++ whoami
101202 INFO DEBUG - + [[ kbin = root ]]
101202 INFO DEBUG - + sudo -u kbin ynh_install_composer
101202 INFO WARNING - sudo: ynh_install_composer: command not found
```

csolisr committed:
fix: remove ynh_exec_as from ynh_install_composer

csolisr commented:
OK I think I have an idea of what happened. `ynh_install_composer` does *not* require the `ynh_exec_as`, but the rest of the commands do. !testme

csolisr commented:
!testme

csolisr commented:
Never mind, the Composer commands also fail here. Any clues on how to make it run as the app user instead of root, while correctly escaping the YNH macros?

csolisr committed:
fix: add experimental helper from Flarum

Added an experimental helper from Flarum, which adds support for
setting the user one wants to run Composer as. This helper was
unfortunately deprecated when ynh_composer_as_user was officially added
to YunoHost, and the specific patch that added user support was not
migrated alongside with this. The patch is taken (and slightly
modified) from the following commit:

https://github.com/YunoHost-Apps/flarum_ynh/blob/0728ea096b4a0cd030b311821e659519b1346f4a/scripts/install

csolisr commented:
I had to go and browse an old version of `flarum_ynh`'s repository to get a version of `ynh_install_composer` that did allow to set the user that would run the application. So, I had to manually import it into the repository.

csolisr commented:
!testme

csolisr committed:
fix: solve linter warning, remove typo

csolisr commented:
!testme

csolisr committed:
fix: correct function names

csolisr commented:
!testme

csolisr commented:
So we still have to bind `$final_path` - how did this helper even work in the first place?

csolisr committed:
chore: update references to old bindings

csolisr commented:
!testme

csolisr committed:
fix: add missing reference to curl

csolisr commented:
!testme

csolisr committed:
fix: correct Composer initial installation logic

csolisr commented:
!testme

csolisr committed:
fix: fix install_args

csolisr commented:
!testme

csolisr committed:
fix: correct parameter name

csolisr commented:
!testme

csolisr committed:
chore: bump version

csolisr commented:
!testme

csolisr committed:
chore: fix checksum

csolisr commented:
!testme

csolisr committed:
fix: add missing dependency for Composer

csolisr commented:
!testme

csolisr commented:
!testme

csolisr committed:
fix: correct Composer installer

ericgaspar committed:
Update install

ericgaspar committed:
Update install

ericgaspar committed:
Update install

ericgaspar committed:
Update install

csolisr committed:
fix: remove quiet parameter temporarily to debug Composer installer

csolisr commented:
Bingo:

```
WARNING Invalid argument dump-env prod. Use "composer require dump-env prod" instead to add packages to your composer.json.
```

csolisr committed:
fix: correct order of operations for Composer

adcura commented on scripts/install:
```diff
@@ -7,175 +7,95 @@
+
+#ynh_install_composer_as_user --phpversion=$phpversion --install_args="--prefer-dist --no-autoloader --no-scripts --no-progress"
+#ynh_install_composer_as_user --phpversion=$phpversion --install_args="dump-env prod"
+ynh_install_composer_asuser --phpversion=$phpversion
```
I'm testing this for fun, in case you're interested, the install failed here at ynh_install_composer_asuser --phpversion=$phpversion with "command not found", I notice that the lines above have the command spelled with an underscore like "...as_user", perhaps that's the reason?  I'll try to figure out a way to modify that line and run the installer again.  Thanks for your efforts so far.

csolisr replied:
Hey you're right about that one! It had somehow slipped out of my eyes, gonna try correcting that line at home

csolisr committed:
fix: typo

csolisr commented:
!testme

csolisr committed:
fix: bind dangling variable

csolisr committed:
fix: another typo

csolisr commented:
To confirm - is this pull request still in progress? @ericgaspar

ericgaspar commented:
!testme

ericgaspar committed:
Update manifest.toml

yunohost-bot committed:
Auto-update README

ericgaspar committed:
cleaning

ericgaspar committed:
Update manifest.toml

ericgaspar commented:
!testme

ericgaspar committed:
cleaning

OniriCorpe committed:
add autoupdate strategy
